src/Vlingo.Actors/Plugin/Mailbox/AgronaMPSCArrayQueue/ManyToOneConcurrentArrayQueueDispatcher.cs

```diff
@@ -32,27 +33,36 @@ public class ManyToOneConcurrentArrayQueueDispatcher : IRunnable, IDispatcher
             Mailbox = new ManyToOneConcurrentArrayQueueMailbox(this, mailboxSize, totalSendRetries);
             this.throttlingCount = throttlingCount;
             closed = new AtomicBoolean(false);
-            cancellationTokenSource = new CancellationTokenSource();
+            dispatcherTokenSource = new CancellationTokenSource();
+            backoffTokenSource = CancellationTokenSource.CreateLinkedTokenSource(dispatcherTokenSource.Token);
         }
 
-        public void Close() => closed.Set(true);
+        public void Close()
+        {
+            closed.Set(true);
+            dispatcherTokenSource.Cancel();
+            dispatcherTokenSource.Dispose();
+            backoffTokenSource.Dispose();
+        }
 
-        public bool IsClosed => closed.Get() || cancellationTokenSource.IsCancellationRequested;
+        public bool IsClosed => closed.Get();
 
         public void Execute(IMailbox mailbox)
         {
-            cancellationTokenSource.Cancel();
+            backoffTokenSource.Cancel();
+            backoffTokenSource.Dispose();
+            backoffTokenSource = CancellationTokenSource.CreateLinkedTokenSource(dispatcherTokenSource.Token);
```

Comment:
Here is an interesting scenario- by the time your code reaches this line, the Thread/Task in the `Run()` method may have already started running and checked that it is "Not Closed" and it "Didn't Deliver" anything. At that moment, it will try to backoff using a canceled and disposed token source. I cannot guarantee that it will happen, neither can I guarantee that it will never happen. This seems to me like a race condition. Or, am I missing something?

Reply:
Nice analysis. However the backoff gets a copy of the token which is a struct, so accessing it is safe even if the owning cancellationtokensource has been cancelled/disposed by the time. Np here.